server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1416,7 +1418,57 @@ public void openEngineAndSkipTranslogRecovery() throws IOException {
         getEngine().skipTranslogRecovery();
     }
 
-    private void innerOpenEngineAndTranslog() throws IOException {
+    /**
+     * Recovers a replica locally up to the global checkpoint before starting a peer recovery.
+     */
+    public void prepareShardForPeerRecovery() {
+        if (state != IndexShardState.RECOVERING) {
+            throw new IndexShardNotRecoveringException(shardId, state);
+        }
+        if (recoveryState.getRecoverySource() instanceof RecoverySource.PeerRecoverySource == false) {
+            assert false : "only peer recovery needs to recover locally up to the global checkpoint";
+            throw new IllegalStateException("only peer recovery needs to recover locally up to the global checkpoint");
+        }
+        try {
+            final long globalCheckpoint;
+            final Engine engine;
+            synchronized (mutex) {
+                String translogUUID = store.readLastCommittedSegmentsInfo().getUserData().get(Translog.TRANSLOG_UUID_KEY);
+                globalCheckpoint = Translog.readGlobalCheckpoint(translogConfig.getTranslogPath(), translogUUID);
+                List<IndexCommit> commits = DirectoryReader.listCommits(store.directory());
+                IndexCommit safeCommit = CombinedDeletionPolicy.findSafeCommitPoint(commits, globalCheckpoint);
+                SequenceNumbers.CommitInfo commitInfo = SequenceNumbers.loadSeqNoInfoFromLuceneCommit(safeCommit.getUserData().entrySet());
+                boolean shouldRecoverLocally = commitInfo.maxSeqNo <= globalCheckpoint && commitInfo.localCheckpoint < globalCheckpoint;
+                if (shouldRecoverLocally && indexSettings.getIndexMetaData().getState() == IndexMetaData.State.OPEN) {
+                    engine = innerOpenEngineAndTranslog();
+                } else {
+                    engine = null;
+                }
+            }
+            if (engine != null) {
+                try {
+                    indexShardOperationPermits.blockOperations(-1, TimeUnit.SECONDS, () -> {
+                        final Engine.TranslogRecoveryRunner translogRecoveryRunner =
+                            (e, snapshot) -> runTranslogRecovery(e, snapshot, Engine.Operation.Origin.LOCAL_RESET, () -> { });
+                        engine.recoverFromTranslog(translogRecoveryRunner, globalCheckpoint);
+                    });
+                    engine.flush(true, true);
+                } finally {
+                    synchronized (mutex) {
+                        currentEngineReference.compareAndSet(engine, null);
+                        IOUtils.close(engine);
+                    }
+                    recoveryState.setStage(RecoveryState.Stage.INIT);
+                    recoveryState.setStage(RecoveryState.Stage.INDEX);
+                }
+            }
+        } catch (Exception ignored) {
+            // We can hit exception if the store is empty or corrupted or incomplete (due to the abort of the previous peer recovery).
```

Comment:
we should at least debug log this. We could also distinguish between expected and unexpected  cases here and log with even higher log level in those unexpected cases (perhaps even add  an assert false).